client/order/epochqueue.go

```diff
@@ -0,0 +1,190 @@
+	si := sort.Search(len(eq.seedIndex), func(i int) bool {
+		order := eq.orders[eq.seedIndex[i]]
+		return bytes.Compare(order.OrderID[:], oid[:]) > 0
+	})
+	eq.seedIndex = append(eq.seedIndex, order.OrderID{})
+	copy(eq.seedIndex[si+1:], eq.seedIndex[si:])
+	eq.seedIndex[si] = oid
+	eq.seedIndexMtx.Unlock()
+
+	// Update the commitment sort order per order commitments.
+	eq.commitmentIndexMtx.Lock()
+	ci := sort.Search(len(eq.commitmentIndex), func(i int) bool {
+		order := eq.orders[eq.commitmentIndex[i]]
+		return bytes.Compare(order.Commitment[:], note.Commitment[:]) > 0
+	})
+	eq.commitmentIndex = append(eq.commitmentIndex, order.OrderID{})
+	copy(eq.commitmentIndex[ci+1:], eq.commitmentIndex[ci:])
+	eq.commitmentIndex[ci] = oid
+	eq.commitmentIndexMtx.Unlock()
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Size() int {
+	eq.ordersMtx.Lock()
+	size := len(eq.orders)
+	eq.ordersMtx.Unlock()
+	return size
+}
+
+// Exists checks if the provided order id is in the queue.
+func (eq *EpochQueue) Exists(oid order.OrderID) bool {
+	eq.ordersMtx.Lock()
+	_, ok := eq.orders[oid]
+	eq.ordersMtx.Unlock()
+	return ok
+}
+
+// GenerateMatchProof calculates the sorting seed used in order matching as well as the
+// commitment checksum from the provided epoch queue preimages and misses.
+func (eq *EpochQueue) GenerateMatchProof(preimages []order.Preimage, misses []string) (int64, msgjson.Bytes, error) {
+	eq.ordersMtx.Lock()
+	defer eq.ordersMtx.Unlock()
+
+	// Remove all misses.
+	for _, entry := range misses {
+		b, err := hex.DecodeString(entry[:])
+		if err != nil {
+			return 0, nil, fmt.Errorf("decoding error: %v", err)
+		}
+		var oid order.OrderID
+		copy(oid[:], b[:])
+
+		delete(eq.orders, oid)
+
+		// Remove the seed index associated with the miss, preserving
+		// the sort order in the process.
+		eq.seedIndexMtx.Lock()
+		si := sort.Search(len(eq.seedIndex), func(i int) bool { return bytes.Compare(eq.seedIndex[i][:], oid[:]) >= 0 })
+		if si == len(eq.seedIndex) {
+			return 0, nil, fmt.Errorf("expected to find id %s in seed sort index", oid)
+		}
+		if si < len(eq.seedIndex)-1 {
+			copy(eq.seedIndex[si:], eq.seedIndex[si+1:])
+		}
+		eq.seedIndex[len(eq.seedIndex)-1] = order.OrderID{}
+		eq.seedIndex = eq.seedIndex[:len(eq.seedIndex)-1]
+		eq.seedIndexMtx.Unlock()
+
+		// Remove the commitment index associated with the miss, preserving
+		// the sort order in the process.
+		eq.commitmentIndexMtx.Lock()
+		ci := sort.Search(len(eq.commitmentIndex), func(i int) bool { return bytes.Compare(eq.commitmentIndex[i][:], oid[:]) >= 0 })
+		if ci == len(eq.commitmentIndex) {
+			return 0, nil, fmt.Errorf("expected to find id %s in commitment sort index", oid)
+		}
+		if ci < len(eq.commitmentIndex)-1 {
+			copy(eq.commitmentIndex[ci:], eq.commitmentIndex[ci+1:])
+		}
+		eq.commitmentIndex[len(eq.commitmentIndex)-1] = order.OrderID{}
+		eq.commitmentIndex = eq.commitmentIndex[:len(eq.commitmentIndex)-1]
```

Comment:
I wonder if it's worth the complexity to keep a sorted list.  It only needs to be sorted in `GenerateMatchProof`, not any other time between calls to `Queue` as far as I can tell, so the order ids and commitments could be kept in maps, which would facilitate these Searches, and for the seed and csum computation, we could form slices and sort them since they are only needed in this function.

Reply:
I think it's worth the trouble keeping track of the sort order for the epoch queue, `GenerateMatchProof` without needing to perform an order id sort and a commitment sort before generating the match proofs would be a lot faster. I think the current implementation would perform substantially better with larger epoch queues.

Reply:
I think this deserves a bit more discussion.  There are several things to consider here:

1. Design goals: Unlike the order book, where keeping the list sorted on each new order received actually is a requirement, epoch validation does not need sorted lists until the epoch close. We do not get this for free; we pay in complexity and performance.

2. Complexity: Performance differences and performance impact (two distinct factors) have to be quite significant to justify additional complexity, which brings bugs and maintenance challenges. A small difference in performance does not justify a notable increase in complexity, and a large performance difference does not justify it either if both solutions have *acceptable* performance (e.g. 10ms vs 20ms in this case I might argue, although we should characterize this).

3. Scalability: The complexity vs. performance trade-off above is much harder to justify for the client, whereas server-side performance is a great concern.  The client has far less to worry about in terms of performance than the server, which has many more roles (e.g. swap coordination and numerous markets), may of which are time critical.

4. Performance: The statement that maintaining sorted lists on each new order insertion "would perform substantially better with larger epoch queues" is not well qualified, and, I believe, is incorrect.

To keep a list sorted, receiving N orders over the course of an epoch involves:
 - N slice location searches, which involves order ID comparisons in a binary search.
 - N copies of up to the entire slice to insert (with the slice growing from length from 0 to N over the epoch).
 
 To sort at epoch close:
 - N appends (no comparisons, and no copies except for occasional reallocations, which the insertions also require in addition to many copies on EACH insertion).
 - 1 sort on epoch close (swaps, no reallocations).  You will find that this is extremely fast if we are sorting a slice of pointers or short arrays.

I intend to benchmark a simple append-and-later-sort approach to compare with the currently proposed solution, and I encourage you to do the same.  But, there would have to be a considerable (and surprising to me) difference in performance to justify the increase in complexity and all that goes with it.

Reply:
Alright, your perspective on the issue is a lot clearer to me now. I'll implement the alternative and benchmark as suggested.

Reply:
Looks like there isn't much being gained from maintaining sort orders on enqueue. `BenchmarkMatchProofAlt`, which generates the sort orders when `GenerateMatchProof` is called is performing at about the same level as `BenchmarkMatchProof` which maintains sort orders. In the current `BenchmarkMatchProofAlt` I'm copying the order id and epoch orders from the order map into slices and sorting that using the order ids and order commitments. 

```
BenchmarkMatchProofAlt500-8           27          41963687 ns/op          414722 B/op       2556 allocs/op
BenchmarkMatchProofAlt1000-8           7         162401527 ns/op          827794 B/op       5064 allocs/op
BenchmarkMatchProofAlt5000-8           1        4149027722 ns/op         4740384 B/op      25170 allocs/op
BenchmarkMatchProof500-8              30          43055826 ns/op          393997 B/op       2549 allocs/op
BenchmarkMatchProof1000-8              7         160078052 ns/op          787010 B/op       5058 allocs/op
BenchmarkMatchProof5000-8              1        4092102140 ns/op         4538520 B/op      25172 allocs/op
```
`BenchmarkMatchProofAlt` does indeed has a simpler implementation, should be preferred as a result. Should I go ahead and make modifications or should I push the pr with the benchmarks?

Reply:
Yes, I think a PR with the benchmarks would be interesting.

Reply:
I was doing some testing and discovered that searching for an `OrderID` in `eq.commitmentIndex` does not work with `sort.Search` because the slice is not sorted by `OrderID`.

```go
ci := sort.Search(len(eq.commitmentIndex), func(i int) bool { return bytes.Compare(eq.commitmentIndex[i][:], oid[:]) >= 0 })
```
 can fail to find `oid`.

Reply:
With a single map and sorting in GenerateMatchProof, here's what I get for a inserting 500/1000/5000 orders and getting the match proof with N/20 misses:

```
BenchmarkEpochQueue500-32     	33	  34366184 ns/op	   298102 B/op	    60 allocs/op
BenchmarkEpochQueue1000-32    	 8      135353264 ns/op	   593462 B/op	    84 allocs/op
BenchmarkEpochQueue5000-32    	 1	3331479770 ns/op	  2454240 B/op	   203 allocs/op
```

I couldn't do the same because of the `commitmentIndex` issue I mentioned in my previous comment.

Reply:
Hit the issue you mentioned with the commitmentIndex, not sure why I thought I could sort the orders with their commitments and index the order ids instead. I'll need to lookup the the orders associated with indexed commitments to maintain the sorting. That should make things slower for the implementation. It's clear to me the complexity involved with maintaining the sorts doesn't provide the performance benefits I thought it would. I'm stripping it out.